### PR TITLE
WinUI NavigationView

### LINKED ISSUE #4317: Hierarchical NavigationView
Add support for the Hierachical NavigationView

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): fixes #4317 

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

Not supported.

## What is the new behavior?

🚀

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

### CODE REVIEW

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+			// Element is already of expected type, just return it
+			if (newContent is NavigationViewItemBase newItem)
+			{
+				return newItem;
+			}
+
+			// Accidentally adding a OS XAML NavigationViewItem to WinUI's NavigationView can cause unnecessary confusion for developers
+			// due to unexpected rendering, potentially without an easy way to understand what went wrong here. To help out developers,
+			// we are explicitly checking for this scenario here and throw a helpful error message so that they can quickly fix their app.
+			if (newContent is Windows.UI.Xaml.Controls.NavigationViewItemBase)
```
Comment (davidjohnoliver) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs:71]: just make sure the type here doesn't get converted by the Uno.WinUI transformation :p

File: src/Uno.UI/UI/Xaml/GridLength.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ public partial struct GridLength : IEquatable<GridLength>
 
 		public bool IsStar { get { return GridUnitType == Xaml.GridUnitType.Star; } }
 
-		public double Value { get; private set; }
+		public double Value { get; internal set; }
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/GridLength.cs:22]: ?

**Review by davidjohnoliver — APPROVED** [review]

File: src/Uno.UI/UI/Xaml/GridLength.cs (hunk 2 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/GridLength.cs:22]: Based on this:

https://github.com/microsoft/microsoft-ui-xaml/blob/bd98a75dabfe0ad0a8956834ea9da735781506e6/dev/NavigationView/NavigationView.cpp#L1756-L1758

But maybe can just create a new `GridLength`.

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs:71]: How do I do that though 😅

File: src/Uno.UI/UI/Xaml/GridLength.cs (hunk 2 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/GridLength.cs:22]: Oh! Fair enough.

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/GridLength.cs]: Changed it anyway, seems not worth introducing mutability :-D

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs:72]: Worked around :-)

Comment (davidjohnoliver) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/NavigationView/NavigationViewItemsFactory.cs:72]: in other contexts the 'fix'seems to be to break the string-matching without changing the semantics, so something like 

```csharp
Windows/*Don't convert this*/.UI.Xaml.Controls.NavigationViewItemBase
```

**Pull request merged**